### ISSUE TITLE
API

### ISSUE THREAD

**Issue by hoyes** [camdram/camdram#24]
Allow other entities to interface with camdram programatically.

Most of the leg-work for this has already been done by using the FOSRestBundle, but remaining issues are:
- Authentication (e.g. implement OAuth using FOSOAuthServerBundle), either as an individual user or group, such that the client obtains the same ACL rights as the corresponding user/group (for creating/deleting shows etc).
  - Interface to allow users/groups to create API keys
  - Documentation.

The following two use cases should be supported:
- Web-based login, e.g., allow visitors to approved websites to 'login using camdram', so that they can take advantage of their camdram profile (shows they're an admin for etc).
- Server-based login, e.g. allow approved websites/venues to pull details of their shows directly from camdram instead of maintaining a separate database

**Comment by hoyes** [comment]
The work above is all now on www.

Still todo:
 - [ ] Show data "roles" url doesn't work - ideally need to rework show routes so that "roles" is the roles page and the thing that's currently using the "roles" url is something else
 - [ ] "Applications" output breaks for society-only applications
 - [ ] Tests to at least smoke test all the API endpoints
 - [ ] Documentation and examples
 - [ ] Come up with a solution for changing slugs
 - [ ] A general once over to check for naming consistency etc
And probably some other things too which I'll add to this as I think of them

**Comment by hoyes** [comment]
Re slug / unique ids:

At the moment, slugs can change at any time, which makes them a bit useless as unique ids. However, Camdram doesn't allow querying by the numeric ID so that's not very useful either. Somehow this needs to be resolved before too many people use the API.

There are also probably implications for SEO / broken external links - at the moment if you modify a show's name then its URL changes.

I wondered whether the below would work as a solution:
 * Give each show, person, society and venue something like a UUID, the first letter of which denotes the type of object (to make it a unique ID across Camdram). UUIDs seem slightly nicer than simply reverted to using the auto-incrementing PK. E.g.:
   * http://www.camdram.net/shows/s8d2o1
   * http://www.camdram.net/people/p0j2id0w
   * http://www.camdram.net/venues/p82dhkd
   * http://www.camdram.net/societies/y2sfhdj3
 * Allow shows to assign themselves a separate, "publicly URL" (can't think of a better name). This could either be fixed at a value something like the current slug (e.g. "2015-rent"), or completely customisable by the user on a first-come-first-served basis. The show's admins would effectively be responsible for picking this and dealing with the consequences of changing it.
 * The "publicty URL" would simply redirect to the UUID-based URL above
 * For reasons of SEO/stopping external links breaking, all existing shows would be given one of these publicly-visible URLs.
 * All new shows wouldn't have a "publicty URL" by default.

The above is ripped off how the BBC do it (e.g. if you go to www.bbc.co.uk/doctorwho, it redirects to www.bbc.co.uk/programmes/b006q2x0). They've clearly thought about it: http://smethur.st/posts/176135860 (see section "Back to URLs").

**Comment by rob-eager** [comment]
Just wanted to say the new show/people.json is beautiful @hoyes! I really appreciate it :grin:

A while ago I wrote a script that generates and formats previous show credits for programme bios (the bane of a programme designer’s life!). It had to use screen scraping but I can now rewrite it to use the API.

As I’d be doing some rewriting I wondered if it would be handy to get it into a form where it could be hosted like the sign in sheet generator so it could be useful to anyone making a programme?

**Comment by stumo** [comment]
Sounds like a great idea. You could either continue to use the public api,
or build it into camdram directly and we link to it on the show page.

**Comment by hoyes** [comment]
Cheers Rob :-)

Yes we could certainly look at incorporating this somehow.

Where does this live at the moment (code / does it have a public URL)? A first step while we figure out what to do might be to simply import whatever you have now into a new repo in the "camdram" Github repo if you'd be happy to do that (we'd obviously give you commit access to that repo).

**Comment by rob-eager** [comment]
Currently it's a set of python classes and scripts that I just run locally probably over half of it is there to do the screen scraping so it might well be easier to start afresh - I imagine it would need to be in php to become a part of the camdram build? - but I'm certainly happy to put it in a repo as a starting point.

**Comment by alexbrett** [comment]
Re: slugs / ids etc - having some sort of unique ID sounds sensible, and I like the idea of having a 'nice' URL that people can publish etc. Something we'd have to be careful to do is to ensure that they are trivially separable, so e.g. ensuring that a unique ID will always start with a letter, and a 'nice' URL will always start with the year.

I'd suggest auto generating the 'nice' URL in the way we currently generate the slug, and showing at least to the show admins what it is so they can copy paste it into e-mails and the like if they don't want to use the UUID one. If you leave it as something people can request, we're likely to get people being silly and having e.g. obscenities and the like in their nice URLs, so auto generating it sounds sensible to me (perhaps with something so a camdram admin can override the auto generated if necessary).

**Comment by hoyes** [comment]
@rob-eager (oops we've got 2 interleaved conversations) 

OK cool. Hmm I think it would have to be PHP in order to be part of Camdram proper. Depend on how complex it would be to translate, an alternate might be to bolt Django/Flask to it and host it on a different URL temporarily. 

If you're able to put in a repo somewhere (even under your username) and stick a link here then we could perhaps take a look and see how much effort it would take to bake it into Camdram.

**Comment by stumo** [comment]
In all honesty, how often are shows renamed? Or people for that matter?

Possibly we should
(a) make slugs static, but with an option to regenerate
(b) keep a slug history so that old slugs become a permanent redirect to
their new one
(c) make it easier for api users to find and use numeric ids (is that
necessary? Works ok for the adc for now)

**Comment by hoyes** [comment]
According to our audit logs for 2015 up until now: 86 times for shows and 54 for people, which I'd argue is enough to make it matter.

We could add something like "/shows/by-id/1234.json", which redirects to a slug-based URL, but then the "canonical" and "redirected" URL would be different (equally if the old slug redirects to the new slug). And it's confusing for the API users - if you use to /shows/asdf.json then it still works so it's tempting to use that as it's the obvious thing to do and it works 99% of the time; if you make /shows/by-id/1234.json output directly without redirecting then you then have 2 different URLs for the same object.

My suspicion is that most users would notice but wouldn't care all that much. (E.g. if you dump the URL into a Facebook post, it expands the page title and description but not the full URL). And it feels nicer and easily to implement than keeping a slug history.

**Comment by philosophicles** [comment]
Another angle here - WHY are people renaming shows so often?

Could this be mostly eliminated, perhaps, by making clearer on the show
creation page, what the resulting url will be, before they hit save? (If
people are tweaking the name to improve the url - though this is also an
argument in favour of decoupling the two things as @hoyes suggests.)

If people initially enter long show names, then see the url and decide to
shorten the show name, perhaps we should offer one off url adjustments at
the point of creation (not after)? Eg if the show name implies this default
slug:

2015-macbeth-tour-of-africa

They could choose to truncate it sooner (2015-macbeth, or
2015-macbeth-tour-of if they wanted), getting a -1 suffix if necessary, but
not adding extra characters.

And / or, re-emphasising that whatever they create will be mostly
unchangeable, so people spellcheck etc rather than creating the show
listing in a hurry?

The spellcheck thing probably accounts for the vast majority of people
renames, I suspect...

**Comment by CHTJonas** [comment]
Conversation is a little hard to follow here. The first parts seem to be about creating the API whilst the last parts seem to relate to #281. @hoyes I think the main intentions from the first part of this issue have been address (save #423) so are you ok to close this? You perhaps hint at API functionality with static auth keys (ie. an application that can make read/write changes to Camdram as a single user/society/venue **without** needing a user to authenticate via OAuth each time) which could maybe be a new issue in & of itself?

@rob-eager are you still interested in incorporating this into Camdram (I know it's been several years so understandable if you've lost your scripts by this point!)

**Comment by hoyes** [comment]
Aye this gone way off topic and I'm at a bit of loss figuring exactly what I meant in the ticket description.

I'll create a ticket for server-side auth. I *think* grant_type=client_credentials will work with the key and secret directly, but have never tested it so would need testing.
